store/src/db/mod.rs

```diff
@@ -174,4 +181,67 @@ impl Db {
 
         Ok(())
     }
+
+    // HELPERS
+    // ---------------------------------------------------------------------------------------------
+
+    /// If the database is empty, generates and stores the genesis block. Otherwise, it ensures that the
+    /// genesis block in the database is consistent with the genesis block data in the genesis JSON
+    /// file.
+    async fn ensure_genesis_block(
+        &self,
+        genesis_filepath: &str,
+    ) -> Result<(), anyhow::Error> {
+        let (expected_genesis_header, account_smt) = {
+            let file_contents = fs::read_to_string(genesis_filepath)?;
+
+            let genesis_state: GenesisState = serde_json::from_str(&file_contents)?;
+            let (block_header, account_smt) = genesis_state.into_block_parts()?;
+
+            (block_header.into(), account_smt)
+        };
+
+        let maybe_block_header_in_store = self.select_block_header_by_block_num(Some(0)).await?;
+
+        match maybe_block_header_in_store {
+            Some(block_header_in_store) => {
+                // ensure that expected header is what's also in the store
+                if expected_genesis_header != block_header_in_store {
+                    return Err(anyhow!("block header in store doesn't match block header in genesis file \"{genesis_filepath}\". expected {expected_genesis_header:?}, but store contained {block_header_in_store:?}"));
```

Comment:
Should this return a more structured error? (e.g., something containing both block headers) This way, we can handle printing this out in a different place.

Reply:
I introduced `GenesisBlockError`, which `ensure_genesis_block()` returns. It gets a bit messy because other store methods that it uses return an `anyhow` error (so I'm forced to introduce a `map_err()` which is distracting when parsing the core logic).

IMO either you use `anyhow` everywhere, or nowhere. I think the store should use structured errors (exactly to be able to define printing logic elsewhere, and have consistent error messages). This is a starting point!

Reply:
Does this mean we should move from `anyhow` to `thiserror`? I guess this is kind of covered in #62, but maybe worth mentioning it there.

Reply:
Yes - I think binaries should use `anyhow`, while libraries should use `thiserror`